internal/kms/azure_vault.go

```diff
@@ -0,0 +1,230 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kms
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/ceph/ceph-csi/internal/util/k8s"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/azure-sdk-for-go/sdk/security/keyvault/azsecrets"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	kmsTypeAzure = "azure-kv"
+
+	// azureDefaultSecretsName is the default name of the Kubernetes Secret
+	// that contains the credentials to access the Azure key vault. The name
+	// the Secret can be configured by setting the `AZURE_CERT_SECRET_NAME` option.
+	//
+	// #nosec:G101, value not credentials, just references token.
+	azureDefaultSecretsName = "ceph-csi-azure-credentials"
+
+	// azureSecretNameKey contains the name of the Kubernetes Secret that has
+	// the credentials to access the Azure key vault.
+	//
+	// #nosec:G101, no hardcoded secret, this is a configuration key.
+	azureSecretNameKey = "AZURE_CERT_SECRET_NAME"
+
+	azureVaultURL = "AZURE_VAULT_URL"
+	azureClientID = "AZURE_CLIENT_ID"
+	azureTenantID = "AZURE_TENANT_ID"
+
+	// The following options are part of the Kubernetes Secrets.
+	//
+	// #nosec:G101, no hardcoded secrets, only configuration keys.
+	azureClientCertificate = "AZURE_CLIENT_CERTIFICATE"
+)
+
+var _ = RegisterProvider(Provider{
+	UniqueID:    kmsTypeAzure,
+	Initializer: initAzureKeyVaultKMS,
+})
+
+type azureKeyVaultKMS struct {
+	// basic
+	namespace  string
+	secretName string
+
+	integratedDEK
+
+	// standard
+	vaultURL          string
+	clientID          string
+	tenantID          string
+	clientCertificate string
+}
+
+func initAzureKeyVaultKMS(args ProviderInitArgs) (EncryptionKMS, error) {
+	kms := &azureKeyVaultKMS{
+		namespace: args.Namespace,
+	}
+
+	// required options for further configuration (getting secrets)
+	err := setConfigString(&kms.secretName, args.Config, azureSecretNameKey)
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	} else if errors.Is(err, errConfigOptionMissing) {
+		kms.secretName = azureDefaultSecretsName
+	}
+
+	err = setConfigString(&kms.vaultURL, args.Config, azureVaultURL)
+	if err != nil {
+		return nil, err
+	}
+	err = setConfigString(&kms.clientID, args.Config, azureClientID)
+	if err != nil {
+		return nil, err
+	}
+	err = setConfigString(&kms.tenantID, args.Config, azureTenantID)
+	if err != nil {
+		return nil, err
+	}
+
+	// read the kubernetes secret with credentials
+	secrets, err := kms.getSecrets()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get secrets for %T, %w", kms, err)
+	}
+
+	err = setConfigString(&kms.clientCertificate, secrets, azureClientCertificate)
+	if err != nil {
+		return nil, err
+	}
+
+	return kms, nil
+}
+
+func (kms *azureKeyVaultKMS) Destroy() {
+	// Nothing to do.
+}
+
+func (kms *azureKeyVaultKMS) getService() (*azsecrets.Client, error) {
```

Comment:
is the client in a connected state that needs disconnecting/destroying once it is not used anymoe?

Reply:
I don't think so, didn't come across Close()/Destory() method for the client object.